language/stdlib/modules/libra_configs.move

```diff
@@ -1,36 +1,92 @@
 address 0x0:
 module LibraConfig {
     use 0x0::Transaction;
-    use 0x0::LibraSystem;
+    use 0x0::LibraAccount;
+    use 0x0::LibraTimestamp;
 
     // A generic singleton resource that holds a value of a specific type.
     resource struct T<Config: copyable> { payload: Config }
 
-    // Set a config item to a new value and trigger a reconfiguration.
-    // This can only be invoked by the Association address
-    public fun set<Config: copyable>(payload: Config) acquires T {
-        let sender = Transaction::sender();
+    struct NewEpochEvent {
+        epoch: u64,
+    }
+
+    resource struct Configuration {
+        epoch: u64,
+        last_reconfiguration_time: u64,
+        events: LibraAccount::EventHandle<NewEpochEvent>,
+    }
 
+    // This can only be invoked by the Association address, and only a single time.
+    // Currently, it is invoked in the genesis transaction
+    public fun initialize_configuration() {
+        let sender = Transaction::sender();
         // Only callable by the Association address for now.
         Transaction::assert(sender == 0xA550C18, 1);
 
-        Transaction::assert(::exists<T<Config>>(sender), 24);
-        let config = borrow_global_mut<T<Config>>(sender);
+        move_to_sender<Configuration>(Configuration {
+            epoch: 0,
+            last_reconfiguration_time: 0,
+            events: LibraAccount::new_event_handle<NewEpochEvent>(),
+        });
+    }
+
+    // We can't return reference here, this returns a copy of the config value.
+    public fun get<Config: copyable>(addr: address): Config acquires T {
+        Transaction::assert(::exists<T<Config>>(addr), 24);
+        let t_ref = borrow_global<T<Config>>(addr);
+
+        *&t_ref.payload
+    }
+
+    // Set a config item to a new value and trigger a reconfiguration.
+    // This can only be invoked by the Association address
+    public fun set<Config: copyable>(addr: address, payload: Config) acquires T, Configuration {
```

Comment:
TODO to add a permission check here? Otherwise anyone can update a config...

Reply:
unfortunately the permission check can not be simply assert(sender == association_address), in the validator admission case, we initialize it in address 0x1D8, but 0xA550C18 can add/remove it, also any current validators can rotate their keys.

discussed with @runtian-zhou yesterday, we think any Config constructor should be gated + some general privilege delegation in libra_config, but also wonder what's your opinion?

Reply:
Sounds good to me. I just want to make sure we don't forget where we need to add checks (I feel less strongly about whether it's implemented via a capability or an `assert`)

Reply:
The pattern in my mind is that we will have a `SetConfigCapability` in this module and a `ValidatorChangeCapability` that utilizes `association.move` 's capability model. We can have `0xa550c18` have `SetConfigCapability` and `0x1D8` have `SetConfigCapability` and `ValidatorChangeCapability`. That way we can enforce special (more privileged) operation for changing validator set.

Reply:
Are using get / set mandatory? Could I directly get a mutable reference copy in a function and modify it? Or even a reference and make smaller copies of it?

If this isn't protected on testnet, what prevents someone from doing something bad?